cmd/kubeadm/app/phases/controlplane/manifests.go

```diff
@@ -228,6 +229,48 @@ func getAPIServerCommand(cfg *kubeadmapi.MasterConfiguration, k8sVersion *versio
 	return command
 }
 
+// calcNodeCidrSize determines the size of the subnets used on each node, based
+// on the pod subnet provided.  For IPv4, we assume that the pod subnet will
+// be /16 and use /24. For IPv6, there are some restrictions. First, the pod
+// subnet size must be /66 or larger. Second, the bit mask used for allocating
+// subnets (one per node) allows up to 32K (16 bits). To accomodate this, the
+// following algorithm is used, based on pod subnet size (implying the number of
+// bits available), to determine the node CIDR size.
+//
+// POD SUBNET SIZE   BITS AVAIL   ACTION
+//    /66  to /104   62 to 24     Allow 16 bits for node CIDR, rest for pods/node
+//    /105 to /112   16 to 23     Give node CIDR half of the bits available
+//    /113 to /119   15 to  9     Give 8 bits for pods/node and rest for node CIDR
+//    /120 to /128   --------     Set node CIDR to pod subnet side - will be rejected later
+//
```

Comment:
Algorithm seems to be ok, but in general, I have doubts that somebody will be trying to fit whole datacenter in single /64 (which is considered minimally viable allocation for interface range in most of the routers).
From aesthetic appearance, if we are about to split available bits between node and pod, it will be good to round it to 16bit for pods. This way by just looking at IPv6 address it will be possible to say which node it belongs to. 
e.g. 

```
2001:0DB8:1234:1234:1234:1234:1234:abcd/112 = pod ip
2001:0DB8:1234:1234:1234:1234:1234::/112 = node range.
```

Reply:
Let me try to clarify the context here. This is not for the entire data center. It is for the pod subnets in the Kubernetes, where the cluster is set up with a **configured** pod network CIDR. From that CIDR, subnets will be created for each node.

For example, if the pod network CIDR was configured to be fd00::/112, then we have 16 bits available. With the above algorithm, 8 bits would be reserved for pods per node, and 8 bits for the number of nodes in the cluster.

The algorithm I chose, tries to balance between a minimum number of pods per node (256+), and keep within the limit for maximum network size (32k nodes). It uses three ranges. For small pod network configurations, it uses 256 pods/node. For large pod networks, it sets number of nodes to max (32k), and the remainder to pods/network. In between these two blocks, it just splits the number of available bits, so that we could have more than 256 pods/node.

Another scheme would be to try to maximize the number of pods/node, using multiples of 8 or 16, giving the rest to the number of nodes in the cluster. That does make the algorithm more complex, as we'd want to have at least 2 nodes in the cluster.  For a multiple of 8, algorithm, we'd have mappings...

/66 - /70 use /72 giving 56 bits for pods/node and 2-6 bits nodes
/71 - /78 use /80 giving 48 bits for pods/node and 2-9 bits for nodes
/79 - /86 use /88 giving 40 bits for pods/node and 2-9 bits for nodes
/87 - /94 use /96 giving 32 bits for pods/node and 2-9 bits for nodes
...

Is there desire to maximize the number of pods per node?

The key point in all this, is that the pod network CIDR is a configured value, and it would be to the benefit of the cluster to choose a "good" value, if readability is important.